File: 173-binary-search-tree-iterator/173-binary-search-tree-iterator.rs

// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 

use std::rc::Rc;
use std::cell::RefCell;

impl TreeNode {
  #[inline]
  pub fn inorder_traversal(&self) -> Vec<i32> {
      
      fn trav (node: &TreeNode, stack: &mut Vec<i32>){
          if let Some(ref left) = node.left{
              trav(&left.borrow(), stack);
          }
          
          stack.push(node.val);
          
          if let Some(ref right) = node.right{
              trav(&right.borrow(), stack);
          }
      }
      
      let mut stack = vec![];
     
      trav(self, &mut stack);
      stack
  }
}

struct BSTIterator {
    array: Vec<i32>,
    index: usize
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl BSTIterator {

    fn new(root: Option<Rc<RefCell<TreeNode>>>) -> Self {
        let mut array = vec![];
        if let Some(node) = root{
            let in_order_stack = node.borrow().inorder_traversal();
            array = in_order_stack;
        }
        
        // println!("{:?}", array);
        BSTIterator{
            array: array,
            index: 0
        }
    }
    
    fn next(&mut self) -> i32 {
        let prev  = self.index;
        self.index+=1;
        
        self.array[prev]
    }
    
    fn has_next(&self) -> bool {
        self.index < self.array.len()
    }
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * let obj = BSTIterator::new(root);
 * let ret_1: i32 = obj.next();
 * let ret_2: bool = obj.has_next();
 */

File: 173-binary-search-tree-iterator/173-binary-search-tree-iterator.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class BSTIterator {
    Stack<TreeNode> stack;
    
    public BSTIterator(TreeNode root) {
        stack = new Stack<TreeNode>();
        putAll(root);
    }
    
    public void putAll(TreeNode root){
        TreeNode node = root;
        
        while(node != null){
            // System.out.println(node.val);
            stack.push(node);
            node = node.left;
        }
    }       
    
    public int next() {
        if (stack.size() > 0){
            TreeNode node = stack.pop();
            
            if (node.right != null){
                putAll(node.right);
            }
            
            return node.val;
        }
        
        return -1;
    }
    
    public boolean hasNext() {
        return stack.size() > 0;
    }
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator obj = new BSTIterator(root);
 * int param_1 = obj.next();
 * boolean param_2 = obj.hasNext();
 */